### PR TITLE
[Verification] makes verification node fork-aware

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
Currently, the `fetcher` engine of the verification nodes assumes that there is a one-to-one mapping between pairs of `(execution result id, chunk index)` and `chunk id`s. However, based on the data extracted from testnet (see the linked issue) this assumption is not true upon execution forks. In other words, due to an execution fork, it happens the same chunk id can be referenced by multiple pairs of (execution result id, chunk index).

**Example**: the figure below illustrates this situation with an example, where there is an execution fork over the `Block D` with two conflicting execution results `ExecutionResult D1` and `ExecutionResult D2`. However, **both these execution results have the same chunk at index 1**.  

If both these identical chunks are assigned to the same verification node and get over processing concurrently, only one chunk may get through verification while the other one is discarded (i.e., deduplicated as it has the same chunk id). This poses a reliability issue as the result approvals are based on chunk index and result ids, and hence one execution result of a fork _may go without approval_ on some of its chunks, which can lead to a sealing halt.

Fork-aware verification example (1)

This PR refactors the way that `fetcher` and `requester` engine of verification node identify the assigned chunks. This is an internal refactoring only confined to verificaiton node. Within these engines, an assigned chunk is maintained by pair of `(its result ID, its chunk Index)`. This pair is called a chunk `Locator`. The `Locator` is not a new data structure in the verification node, however, identifying chunks based on their `Locator` id is the main change this PR introduces. Note that with a locator, the chunk and chunk ID is easily retrievable. The uniqueness of chunks base on their chunk id, hence is migrated to their uniqueness of locator ID (only in the aforementioned engines). With these changes, if the execution results of an execution fork share the same chunk (with same ID) (aka duplicate chunk), and more than one of those duplicate chunks are assigned to the verification node, the `fetcher` and `requester` engine presume each of those chunks _distinct_ (by their result ID and chunk index), and hence process them both. 

This PR also adds an optimization that out of duplicate chunk data requests coming to the requester engine, only one chunk data pack is requested, and upon receiving the response, the requester engine addresses all chunk data requests with that response.

### CODE REVIEW

File: engine/verification/fetcher/engine.go [hunk 1]
```diff
@@ -316,9 +315,7 @@ func (e *Engine) handleValidatedChunkDataPack(ctx context.Context,
 	status *verification.ChunkStatus,
 	chunkDataPack *flow.ChunkDataPack) (bool, error) {
 
-	chunk := status.ExecutionResult.Chunks[status.ChunkIndex]
-	removed := e.pendingChunks.Rem(chunkDataPack.ChunkID)
-
+	removed := e.pendingChunks.Rem(status.ChunkIndex, status.ExecutionResult.ID())
```
Comment (smnzhu) [inline engine/verification/fetcher/engine.go:318]: If it wouldn't be too big of a change, could we rename this method to `Remove` while we're at it 😆

File: engine/verification/fetcher/engine.go [hunk 2]
```diff
@@ -361,8 +359,14 @@ func (e *Engine) validateChunkDataPack(chunkIndex uint64,
 	result *flow.ExecutionResult) error {
 
 	chunk := result.Chunks[chunkIndex]
+	// 1. chunk ID of chunk data pack should map the chunk ID on execution result
+	expectedChunkID := result.Chunks[chunkIndex].ID()
```
Comment (smnzhu) [inline engine/verification/fetcher/engine.go:363]:
```suggestion
	expectedChunkID := chunk.ID()
```

File: model/flow/chunk.go [hunk 3]
```diff
@@ -55,7 +55,7 @@ func (c *ChunkDataPack) Checksum() Identifier {
 	return MakeID(c)
 }
 
-// Note that this is the basic version of the List, we need to substitute it with something like Merkel tree at some point
+// TODO: that this is the basic version of the list, we need to substitute it with something like Merkel tree at some point
```
Comment (smnzhu) [inline model/flow/chunk.go:58]:
```suggestion
// TODO: This is the basic version of the list, we need to substitute it with something like Merkle tree at some point
```

File: engine/verification/fetcher/execution_fork_test.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	// mocks resources on fetcher engine side.
+	mockResultsByIDs(s.results, []*flow.ExecutionResult{resultA, resultB})
+	mockBlocksStorage(s.blocks, s.headers, block)
+	mockPendingChunksAdd(t, s.pendingChunks, assignedChunkStatuses, true)
+	mockPendingChunksRem(t, s.pendingChunks, assignedChunkStatuses, true)
+	mockPendingChunksGet(s.pendingChunks, assignedChunkStatuses)
+
+	// fetcher engine must create a chunk data request for each of chunk statusA and statusB
+	requestA := chunkRequestFixture(resultA.ID(), statusA, executorsA, executorsB)
+	requestB := chunkRequestFixture(resultB.ID(), statusB, executorsB, executorsA)
```
Comment (jordanschalm) [inline engine/verification/fetcher/execution_fork_test.go:53]: So in the execution fork case, we send 2 CHDP requests, but only process one of the responses? (see previous comment in `requester.go`)

**Review by jordanschalm — COMMENTED** [review]
Had some questions about the processing of inbound chunk data packs

File: engine/verification/requester/requester.go [hunk 5]
```diff
@@ -191,16 +191,25 @@ func (e *Engine) handleChunkDataPack(originID flow.Identifier, chunkDataPack *fl
 	e.metrics.OnChunkDataPackResponseReceivedFromNetworkByRequester()
 
 	// makes sure we still need this chunk, and we will not process duplicate chunk data packs.
-	removed := e.pendingRequests.Rem(chunkID)
+	locators, removed := e.pendingRequests.PopAll(chunkID)
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply engine/verification/requester/requester.go:194]: > Suppose EN1 and EN2 produce 2 execution results for the same block, which have different IDs and constitute an execution fork. We are processing chunk C.

Let's formulate it more: suppose we have `result1` and `result2` both for the same block, but with different result IDs, hence an execution fork. For each chunk `c1` in `result1` and `c2` in `result2` that _share the same chunk index_ there will be either of the following cases:

- Case 1: `c1` and `c2` are two _distinct_ chunks. Then they have distinct chunk IDs. Hence each of these chunks are verified separately, i.e., a separate chunk data request for `c1` and a separate one for `c2` and then they take their exclusive verification path. This is the existing verification logic on `master` and this PR does not change it. 

- Case 2: `c1` and `c2` are _identical_ chunks. Then they have the same chunk ID. Although each of these chunks again are treated distinctly and have their own verification path, they share their chunk data pack. This follows the assumption that chunk data packs are matched with chunk IDs. Hence, only a single chunk data pack is requested for both of these chunks. Once that chunk data pack arrives, both of these chunks `c1` and `c2` resume their verification path separately. This was not previously supported, and is the core functionality added by this PR. Without this PR in place, if a verification node is assigned both `c1` and `c2` concurrently, it deduplicates one of them in the favor of the other one. However, since each of them belong to a distinct result, and since result approvals are attributed to the execution results, one result would end up having no approval for its chunk. This situation happend on testnet and the PR is fixing that. 

Note that in both cases `c1` and `c2` will have their own result approvals.

File: engine/verification/fetcher/execution_fork_test.go (hunk 4 above)
Comment (yhassanzadeh13) [reply engine/verification/fetcher/execution_fork_test.go:53]: This test only accounts for execution forks with duplicate chunks, e.g., `result1` and `result2` are execution results of the same block, and `result1.ID()` is different than `result2.ID()`. However, both results have the same chunk at their index 0, i.e., `result1.Chunks[0] = result2.Chunks[0]`.

It also assumes that both chunks are assigned to this verification node and are processed concurrently.

For chunk index 0, since both results have the same chunk, and hence the same chunk ID, and since chunk data packs are associated with chunk IDs, we send two chunk requests to the requester engine. However, requester engine that receives these two chunk requests concurrently, realizes that they both belong to the same chunk ID, hence only requests one chunk data pack for that chunk ID. 

When requester engine receives the chunk data pack back, it sends that chunk data pack twice to the fetcher engine: once for `resultA.Chunks[0]` and once for `resultB.Chunks[0]`. The reason is each of these chunks take their own verification path and literally two result approvals will be generated: one for chunk index 0 of `resultA` and the other one for chunk index 0 or `resultB`. 

Note that if these two duplicate chunks are coming sequentially at the non-overlapping intervals, the requester will treat them distinct, i.e., a chunk data pack is requested for each of them. Once a chunk data pack response is delivered to fetcher engine, the requester engine cleans the memory about it, so any further duplicate requests will also be processed. However, this test covers the edge case of concurrently processing duplicate chunks of an execution forks.

File: engine/verification/fetcher/engine.go [hunk 6]
```diff
@@ -289,24 +288,24 @@ func (e *Engine) handleChunkDataPackWithTracing(
 	status *verification.ChunkStatus,
 	chunkDataPack *flow.ChunkDataPack) (bool, error) {
 
-	var ferr error
-	processed := false
-
 	// make sure the chunk data pack is valid
 	err := e.validateChunkDataPackWithTracing(ctx, status.ChunkIndex, originID, chunkDataPack, status.ExecutionResult)
 	if err != nil {
-		// TODO: this can be due to a byzantine behavio
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply engine/verification/fetcher/engine.go:298]: Thanks for catching, added here: https://github.com/onflow/flow-go/pull/1538/commits/06f41cc64fed8384d8a5a0bfd9175487e91e8597

File: engine/verification/fetcher/engine.go [hunk 7]
```diff
@@ -473,9 +479,9 @@ func (e *Engine) NotifyChunkDataPackSealed(chunkID flow.Identifier) {
 	lg = lg.With().
 		Uint64("block_height", status.BlockHeight).
 		Hex("result_id", logging.ID(status.ExecutionResult.ID())).Logger()
-	removed := e.pendingChunks.Rem(chunkID)
+	removed := e.pendingChunks.Rem(chunkIndex, resultID)
 
-	e.chunkConsumerNotifier.Notify(status.ChunkLocatorID())
+	e.chunkConsumerNotifier.Notify(chunks.ChunkLocatorID(resultID, chunkIndex))
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply engine/verification/fetcher/engine.go]: https://github.com/onflow/flow-go/pull/1538/commits/a5031f441b6d07591f9f88115b91450b0ba7b3ce

File: engine/verification/fetcher/engine.go [hunk 8]
```diff
@@ -271,7 +270,7 @@ func (e *Engine) HandleChunkDataPack(originID flow.Identifier, chunkDataPack *fl
 		e.metrics.OnVerifiableChunkSentToVerifier()
 
 		// we need to report that the job has been finished eventually
-		e.chunkConsumerNotifier.Notify(status.ChunkLocatorID())
+		e.chunkConsumerNotifier.Notify(chunks.ChunkLocatorID(resultID, status.ChunkIndex))
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply engine/verification/fetcher/engine.go]: https://github.com/onflow/flow-go/pull/1538/commits/a5031f441b6d07591f9f88115b91450b0ba7b3ce

File: engine/verification/fetcher/engine.go (hunk 1 above)
Comment (yhassanzadeh13) [reply engine/verification/fetcher/engine.go:319]: `Rem` seems a naming convention for `Remove` method across all of our mempools, e.g., [`ExecutionResults` ](https://github.com/onflow/flow-go/blob/master/module/mempool/results.go#L18). Also, almost all these implementations directly invoke [`Backend.Rem`](https://github.com/onflow/flow-go/blob/1e0a30e7a9118aad96a706639e37ec955f562542/module/mempool/stdmap/results.go#L31-L34). Refactoring `Rem` method for `pendingChunks` would be simple. But refactoring it for all would be a larger surface.

**Review by jordanschalm — APPROVED** [review]
Nice work! 🎸 

Summarizing our conversation:

There is a case where we could have two chunks with the same chunk ID, different execution result IDs, and different chunk data packs. Currently we assume chunks with the same ID will have the same chunk data pack. 

How does this case happen?
1. There is an execution fork in chunk `C` of execution result `R`
2. EN1 executes `C` and computes end state `ES1`. EN2 executes `C` and computes end state `ES2`. `ES1 != ES2`. This is the execution fork. We assume EN1 and EN2 read/write different registers, resulting in a different chunk data pack.
3. A VN is assigned to `C` and requests the chunk data pack for `C`. The VN processes both chunks separately but uses the first chunk data pack it receives for both. 

Note: `chunk.ID()` [does not include end state](https://github.com/onflow/flow-go/blob/b51ce0b3dc1391c18e2a08809730c0417609025a/model/flow/chunk.go#L26)

Suggestions to address this in a separate PR:
1. When requesting chunk data packs, request based on the executor of the ER. 
2. Gracefully handle a chunk data pack with missing registers (we need to do this anyway for BFT)

**Review by zhangchiqing — APPROVED** [review]
Nice fix. 

It's better to add some comment to explain with examples the difference between ChunkID and ChunkLocatorID.

ChunkID is unique within a single execution result fork. But if there is multiple execution result forks, the same chunk ID might appear on different execution result forks. Whereas, the ChunkLocatorID is unique across multiple execution result forks.

File: module/mempool/stdmap/chunk_requests.go [hunk 9]
```diff
@@ -73,13 +61,37 @@ func (cs *ChunkRequests) RequestHistory(chunkID flow.Identifier) (uint64, time.T
...
+			status := &chunkRequestStatus{
+				Locators:    chunks.LocatorList{&request.Locator},
+				RequestInfo: request.ChunkDataPackRequestInfo,
+			}
+			backdata[request.ChunkID] = status
+			return nil
+		}
+
+		status := toChunkRequestStatus(entity)
+		if status.Locators.Contains(request.Locator.ResultID, request.Locator.Index) {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply module/mempool/stdmap/chunk_requests.go]: https://github.com/onflow/flow-go/pull/1538/commits/804783f4734eee83a666335bc734125cea155bed

File: module/mempool/stdmap/chunk_statuses.go [hunk 10]
```diff
@@ -67,3 +78,25 @@ func (cs ChunkStatuses) All() []*verification.ChunkStatus {
...
+
+// storedChunkStatus is an internal type for storing ChunkStatus in the mempool.
+//
+// It is the same as ChunkStatus, but additionally it implements an Entity type which
+// makes it storable in the mempool.
+// Note that as an entity, the ID of a storedChunkStatus is computed as the ID of the chunk locator
+// it represents. However, the usage of ID method is only confined to maintaining it on the mempool.
+// That is the motivation behind making it an internal type to make sure that no further decision out of
+// this package is taken based on ID of storedChunkStatus.
+type storedChunkStatus struct {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply module/mempool/stdmap/chunk_statuses.go]: https://github.com/onflow/flow-go/pull/1538/commits/9a9972a63fcfcde3485b5f7190729133ee72d209

File: module/mempool/stdmap/chunk_statuses.go [hunk 11]
```diff
@@ -67,3 +78,25 @@ func (cs ChunkStatuses) All() []*verification.ChunkStatus {
...
+	ChunkIndex      uint64
+	BlockHeight     uint64
+	ExecutionResult *flow.ExecutionResult
+}
+
+func (s storedChunkStatus) ID() flow.Identifier {
+	return chunks.ChunkLocatorID(s.ExecutionResult.ID(), s.ChunkIndex)
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply module/mempool/stdmap/chunk_statuses.go:97]: Currently is used once per storing into memory (`Add`) and also once per adjusting the size of `Backend` if it gets full. Not sure if caching worth it, but can address it in a separate PR if you recommend (so that it will be reversible if needed).
